storage: eagerly carry out chunked GC

tbg opened the pull request:
The GC queue was already splitting up the set of keys to be deleted into
small parts to avoid overloading the Raft machinery. From there, it was
only a small step to let it dispatch the corresponding GCRequests eagerly.

The upshot is that now we don't have to hold all of the keys to be deleted
in memory any more.

With appropriate testing, I think this is a change suitable for inclusion
in 2.0.1. "Appropriate testing" includes a roachtest which creates a large
range (larger than available RAM) and fills it with GC'able data. (A
variant of the `DROP` test that increases the range size suitably should do
it).

Release note (performance improvement): lowered the amount of memory used
during garbage collection of old versions.

tbg committed:
rditer: add TODO about API shortcoming

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24209)
<!-- Reviewable:end -->

bdarnell commented:
> creates a large range (larger than available RAM) 

Shouldn't that be disallowed by the backpressure introduced by #21777?

This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for 2.1 because it is a step towards increasing range size limits, but before we consider this for 2.0.1 we should do more investigation to figure out how we got such an oversized range to begin with. 

---

Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 517 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Ywox9qQzEDFJMadqR:-L8Ywox9qQzEDFJMadqS:bc5gjoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L517)):*
> ```Go
> }
> 
> // NoopGCer implements storage.GCer by doing nothing.
> ```

Nit: we're in package storage so no need to repeat it in the doc comments.

---

*[pkg/storage/gc_queue.go, line 550 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8YxUf0YYaVuMOgIdZ1:-L8YxUf0YYaVuMOgIdZ2:b-f82weu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L550)):*
> ```Go
> 	var ba roachpb.BatchRequest
> 
> 	// Technically not needed since we're talking directly to the Range.
> ```

s/Range/Replica/

---

*[pkg/storage/gc_queue.go, line 568 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Yy2dOy8c0B3h5sbu5:-L8Yy2dOy8c0B3h5sbu6:bkejxs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L568)):*
> ```Go
> 	req.Threshold = thresh.Key
> 	req.TxnSpanGCThreshold = thresh.Txn
> 	return r.send(ctx, req)
> ```

Comment that we specifically do not batch SetGCThreshold with keys to GC to minimize contention in the command queue. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-:-L8Yw9lhFvHGa_IIK3DK:b8sq8pj)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
And approved for 2.0.1 following discussion in #24215

nvanbenschoten commented:
Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 538 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVS3QXiyIRzgUziLS:-L8ZVS3QXiyIRzgUziLT:bna1j8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L538)):*
> ```Go
> 	desc := r.repl.Desc()
> 	var template roachpb.GCRequest
> 	template.Key = desc.StartKey.AsRawKey()
> ```

Consider adding a TODO to constrain this header to only what's actually in the GCRequest. Now that we're chunking these requests, we can be a lot more CommandQueue friendly.

---

*[pkg/storage/gc_queue.go, line 555 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZU3LmSYAHSZ8PrW0A:-L8ZU3Lni6A_3xhLRuiJ:b-xjxro4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L555)):*
> ```Go
> 	var union roachpb.RequestUnion
> 	union.MustSetInner(&req)
> 	ba.Requests = append(ba.Requests, union)
> ```

`ba.Add`?

---

*[pkg/storage/gc_queue.go, line 568 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Yy2dOy8c0B3h5sbu5:-L8ZVlXCh_HyFaldcABO:b-5h7exo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L568)):*
> minimize contention in the command queue

How does this minimize contention? Just because SetGCThreshold will now be very quick? They both still declare a write over the entire range of keys.

On this note, what time do these GCRequests indicate to the CommandQueue that they're writing at? We should be able to move the requests into the past to the threshold timestamp such that https://github.com/cockroachdb/cockroach/pull/14342 ensures that future reads don't block on them. In fact, writes shouldn't need to block on them either, although I think there would be practical complications to allowing that.

---

*[pkg/storage/gc_queue.go, line 819 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZV1SjdpgrNpIIXK_b:-L8ZV1SjdpgrNpIIXK_c:b83x5s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L819)):*
> ```Go
> 							// Allow releasing the memory backing batchGCKeys.
> 							iter.ResetAllocator()
> 							batchGCKeys = []roachpb.GCRequest_GCKey{}
> ```

`batchGCKeys = nil`

---

*[pkg/storage/gc_queue.go, line 836 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVMNQuUXwWUIMgkd6:-L8ZVMNQuUXwWUIMgkd7:b-av2de5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5dc5617f5fc3fc26ac84dd745559b8d1babfa07/pkg/storage/gc_queue.go#L836)):*
> ```Go
> 
> 	// From now on, all newly added keys are range-local.
> 	// TODO(tschottdorf): Might need to use two requests at some point since we
> ```

Does this TODO go away?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Shouldn't that be disallowed by the backpressure introduced by #21777?

True (and exlained in #24215). The comment also applies to scenarios in which we can run with range sizes larger than the available RAM, though, which I hope will be a real option in 2.1 or 2.2 (and which will be what the roachtest will do, except it waits for upreplication before growing the data set to avoid snapshot blow-out).

---

Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/gc_queue.go, line 517 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Ywox9qQzEDFJMadqR:-L8ZgH--3pV4FfBWSMld:b-w84ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L517)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Nit: we're in package storage so no need to repeat it in the doc comments.
</blockquote></details>

Forgot to update this when I moved this to storage, done.

---

*[pkg/storage/gc_queue.go, line 538 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVS3QXiyIRzgUziLS:-L8ZcETan6XGogFradz1:beupjqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L538)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding a TODO to constrain this header to only what's actually in the GCRequest. Now that we're chunking these requests, we can be a lot more CommandQueue friendly.
</blockquote></details>

This is actually happening under the hood already, see https://github.com/cockroachdb/cockroach/blob/fbffeec6dd5e75bd15933e66c9dcf8bb4db0214a/pkg/storage/batcheval/cmd_gc.go#L36. We've argued for a while that we should be sending individual requests instead, see https://github.com/cockroachdb/cockroach/issues/7880, and then we wouldn't have to set this fake span (which we use to bounce the commands on split out of paranoia) at all.
Let me know if you still think there should be a TODO.

---

*[pkg/storage/gc_queue.go, line 550 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8YxUf0YYaVuMOgIdZ1:-L8ZcfDV2twiq3JKK8OL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L550)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/Range/Replica/
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 555 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZU3LmSYAHSZ8PrW0A:-L8ZcjXyxj8e7WWkSiwx:b9c2275) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L555)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`ba.Add`?
</blockquote></details>

:man_facepalming: Done

---

*[pkg/storage/gc_queue.go, line 568 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Yy2dOy8c0B3h5sbu5:-L8ZcnQgUPl5Oo8zaW1B:b-y57cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L568)):*
See the other comment above. GCRequest declares a nonstandard set of keys.

> On this note, what time do these GCRequests indicate to the CommandQueue that they're writing at?

They declare `clock.Now()` (see `template()`). You're right that we could run them at the larger of the two involved thresholds instead. That could actually be a nice win. And yeah, ditto for writes. Could we achieve this by having the requests declare synthetic keys only checked by other GCRequests? I.e. if there is a GCRequest for `/table/a` we really declare `/local/magic/table/a` to achieve that GCRequests will block on each other, but other traffic doesn't. I don't like the magic of that, though.

Either way, we shouldn't introduce any of these things without a benchmark. Some queue-type workload roachtests are overdue anyway, those should be able to profit from these kinds of tweaks.

Filed https://github.com/cockroachdb/cockroach/issues/24237.

---

*[pkg/storage/gc_queue.go, line 819 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZV1SjdpgrNpIIXK_b:-L8Zg4kiWg9JGzFSByMZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L819)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`batchGCKeys = nil`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 836 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVMNQuUXwWUIMgkd6:-L8Zg6GPXR-e9CjIShWt:b-v1pdvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5dc5617f5fc3fc26ac84dd745559b8d1babfa07/pkg/storage/gc_queue.go#L836)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this TODO go away?
</blockquote></details>

I'm not seeing a TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-:-L8ZbUUuHOZmG9ANpfcq:bs6f0ga)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 538 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVS3QXiyIRzgUziLS:-L8cCFd7htAbCboPLoCh:bodgi9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L538)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This is actually happening under the hood already, see https://github.com/cockroachdb/cockroach/blob/fbffeec6dd5e75bd15933e66c9dcf8bb4db0214a/pkg/storage/batcheval/cmd_gc.go#L36. We've argued for a while that we should be sending individual requests instead, see https://github.com/cockroachdb/cockroach/issues/7880, and then we wouldn't have to set this fake span (which we use to bounce the commands on split out of paranoia) at all.
Let me know if you still think there should be a TODO.
</blockquote></details>

Ah ok, I didn't know we were doing that. Thanks for explaining. No need for a TODO.

---

*[pkg/storage/gc_queue.go, line 568 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8Yy2dOy8c0B3h5sbu5:-L8cARt-HvDBtAPqAURB:b-hmanq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L568)):*
Thanks for filing!

> Could we achieve this by having the requests declare synthetic keys only checked by other GCRequests?

That should work, but this has me curious about something else. If GCRequest is declared at `clock.Now()`, is it ignored by historical reads in the commandQueue? In that case, how does the synchronization work when we check that the `requestCanProceed`? Is it possible that a historical read passes this check but still finds that the keys it is trying to read have been GCed?

---

*[pkg/storage/gc_queue.go, line 836 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209#-L8ZVMNQuUXwWUIMgkd6:-L8cAF0K3_AroyIX97o2:b-gealid) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5dc5617f5fc3fc26ac84dd745559b8d1babfa07/pkg/storage/gc_queue.go#L836)):*
The one that was removed. It still seems applicable.

> // TODO(tschottdorf): Might need to use two requests at some point since we
> 	// hard-coded the full non-local key range in the header, but that does
> 	// not take into account the range-local keys. It will be OK as long as
> 	// we send directly to the Replica, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24209)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 568 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24209#-L8Yy2dOy8c0B3h5sbu5:-L9MxCsYCC3q5CNXf-xJ:b6v2tns) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f04be37034f89cbf091f178a190fa89213c21f6/pkg/storage/gc_queue.go#L568)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks for filing!

> Could we achieve this by having the requests declare synthetic keys only checked by other GCRequests?

That should work, but this has me curious about something else. If GCRequest is declared at `clock.Now()`, is it ignored by historical reads in the commandQueue? In that case, how does the synchronization work when we check that the `requestCanProceed`? Is it possible that a historical read passes this check but still finds that the keys it is trying to read have been GCed?
</blockquote></details>

I think you *might* be right and we should fix it but morally this won't happen. Note that we bump the thresholds in the first command that doesn't actually GC any keys. This means that in practice it's visible to all reads by the time we actually remove data. I think there's a scenario in which a read slips by the GC bump but gets stalled before opening a RocksDB snapshot, so that when it finally does we have actually already removed keys? Setting `ReplicatedEvalResult.BlockReads = true` should fix that problem.

Filed https://github.com/cockroachdb/cockroach/issues/new.

---

*[pkg/storage/gc_queue.go, line 836 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24209#-L8ZVMNQuUXwWUIMgkd6:-L9Mzu4hC0MYzhZOQ9_p:b-8qz1wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5dc5617f5fc3fc26ac84dd745559b8d1babfa07/pkg/storage/gc_queue.go#L836)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The one that was removed. It still seems applicable.

> // TODO(tschottdorf): Might need to use two requests at some point since we
> 	// hard-coded the full non-local key range in the header, but that does
> 	// not take into account the range-local keys. It will be OK as long as
> 	// we send directly to the Replica, though.

</blockquote></details>

I'm not sure about that TODO any more. `GCRequest` is a little weird in that its span doesn't mean much as long as it hits the right range. We use the whole range span, so anything that addresses there is also covered.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24209)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: eagerly carry out chunked GC

The GC queue was already splitting up the set of keys to be deleted into
small parts to avoid overloading the Raft machinery. From there, it was
only a small step to let it dispatch the corresponding GCRequests
eagerly.

The upshot is that now we don't have to hold all of the keys to be
deleted in memory any more.

With appropriate testing, I think this is a change suitable for
inclusion in 2.0.1. "Appropriate testing" includes a roachtest which
creates a large range (larger than available RAM) and fills it with
GC'able data. (A variant of the `DROP` test that increases the range
size suitably should do it). Such a test is filed as #24214.

Release note (performance improvement): lowered the amount of memory
used during garbage collection of old versions.

tschottdorf commented:
bors r+
